src/uci.cpp

```diff
@@ -53,18 +49,30 @@ template<typename... Ts>
 overload(Ts...) -> overload<Ts...>;
 
 UCIEngine::UCIEngine(int argc, char** argv) :
-    engine(argv[0], [](const std::string& str) { sync_cout << str << sync_endl; }),
+    engine(argv[0]),
     cli(argc, argv) {
 
+    auto& options = engine.get_options();
+
+    options.add_listener("NumaPolicy", [](const std::optional<std::string>& str) {
```

Comment:
I don't think this listener machinery is needed, or at least should not be used for this basic functionality. The printing of the output the engine *wants* to be printed to the user should be by default.

I'd consider adding a broader listener. For example every option would have a reference to its OptionsMap, and call `OptionsMap::send_message_to_user` (or simply `info`, as in a log) `on_change` is called, with the value returned from `on_change` if it's not std::nullopt. OptionsMap would propagate to listener[s] and UCIEngine would format it as CLI output.

Also, I believe we have to have all such lines starting with `info string`, so UCIEngine needs to split this desired output by line and print line by line.

Reply:
> The printing of the output the engine wants to be printed to the user should be by default.

we wanted to make all cout optionals, which means delegating all couts to the uciengine?

> I'd consider adding a broader listener.

from a consumer perspective i'd find that annoying, i'd have to write additional logic 
just to differentiate events?

> Also, I believe we have to have all such lines starting with info string, so UCIEngine needs to split this desired output by line and print line by line.

yeah, `NumaPolicy` needs to split it and reformat it, overlooked that

Reply:
> we wanted to make all cout optionals, which means delegating all couts to the uciengine?

I meant that the on_change would return the info string optionally, but it should always be printed by the UCIEngine. We should not be printing anywhere else ideally

> from a consumer perspective i'd find that annoying, i'd have to write additional logic
just to differentiate events?

In my mind the consumer is obliged to print (or pass to the user in some way) the thing on_change wants to pass. So it shouldn't be allowed to choose what to listen to, it needs to listen to everything. Besides right now the UCIEngine needs to know what options the Engine is exposing a priori.

Reply:
i think the current state looks more like the approach you described?